File: autocontext/src/autocontext/scenarios/custom/agent_task_spec.py
from __future__ import annotations

import json
from dataclasses import dataclass, replace
from typing import Any


@dataclass(slots=True)
class AgentTaskSpec:
    """Specification for an agent task scenario."""

    task_prompt: str
    judge_rubric: str
    output_format: str = "free_text"  # free_text | json_schema | code
    judge_model: str = ""
    difficulty_tiers: list[dict] | None = None
    reference_context: str | None = None
    reference_sources: list[str] | None = None
    required_concepts: list[str] | None = None
    calibration_examples: list[dict] | None = None
    context_preparation: str | None = None  # Instructions for context gathering
    required_context_keys: list[str] | None = None  # Keys that must be in state after prepare_context
    max_rounds: int = 1  # Max improvement rounds (1 = single-shot)
    quality_threshold: float = 0.9  # Stop improving when score >= this
    revision_prompt: str | None = None  # Instructions for how to revise output
    sample_input: str | None = None  # Sample input data for data-dependent tasks


def _serialize_agent_task_text_payload(value: Any) -> str | None:
    if value is None:
        return None
    if isinstance(value, str):
        return value
    if isinstance(value, dict | list):
        return json.dumps(value, indent=2)
    return str(value)


def normalize_agent_task_runtime_fields(spec: AgentTaskSpec) -> AgentTaskSpec:
    """Coerce structured prompt-adjacent fields into runtime-safe strings.

    LLM-designed agent-task specs occasionally return structured JSON for fields
    like sample_input. The generated runtime embeds those fields into prompts via
    string concatenation, so we normalize them once at the spec boundary.
    """
    return replace(
        spec,
        task_prompt=_serialize_agent_task_text_payload(spec.task_prompt) or "",
        judge_rubric=_serialize_agent_task_text_payload(spec.judge_rubric) or "",
        reference_context=_serialize_agent_task_text_payload(spec.reference_context),
        context_preparation=_serialize_agent_task_text_payload(spec.context_preparation),
        revision_prompt=_serialize_agent_task_text_payload(spec.revision_prompt),
        sample_input=_serialize_agent_task_text_payload(spec.sample_input),
    )

File: ts/src/scenarios/agent-task-spec.ts
/**
 * AgentTaskSpec — specification for an agent task scenario.
 * Port of autocontext/src/autocontext/scenarios/custom/agent_task_spec.py
 */

import { z } from "zod";

export const AgentTaskSpecSchema = z.object({
  taskPrompt: z.string().min(1, "task_prompt must not be empty"),
  judgeRubric: z.string().min(1, "judge_rubric must not be empty"),
  outputFormat: z.enum(["free_text", "json_schema", "code"]).default("free_text"),
  judgeModel: z.string().min(1, "judge_model must not be empty").default("claude-sonnet-4-20250514"),
  difficultyTiers: z.array(z.record(z.unknown())).nullable().optional(),
  referenceContext: z.string().min(1, "reference_context, if provided, must not be empty").nullable().optional(),
  referenceSources: z.array(z.string().min(1)).min(1, "reference_sources, if provided, must not be empty").nullable().optional(),
  requiredConcepts: z.array(z.string().min(1)).min(1, "required_concepts, if provided, must not be empty").nullable().optional(),
  calibrationExamples: z.array(z.record(z.unknown())).nullable().optional(),
  contextPreparation: z.string().min(1, "context_preparation, if provided, must not be empty").nullable().optional(),
  requiredContextKeys: z.array(z.string().min(1)).min(1, "required_context_keys, if provided, must not be empty").nullable().optional(),
  maxRounds: z.number().int().min(1, "max_rounds must be >= 1").default(1),
  qualityThreshold: z.number().gt(0).lte(1, "quality_threshold must be between 0.0 (exclusive) and 1.0 (inclusive)").default(0.9),
  revisionPrompt: z.string().min(1, "revision_prompt, if provided, must not be empty").nullable().optional(),
  sampleInput: z.string().min(1, "sample_input, if provided, must not be empty").nullable().optional(),
});

export type AgentTaskSpec = z.infer<typeof AgentTaskSpecSchema>;

/**
 * Parse a raw JSON object (snake_case from LLM) into an AgentTaskSpec.
 */
export function parseRawSpec(data: Record<string, unknown>): AgentTaskSpec {
  return AgentTaskSpecSchema.parse({
    taskPrompt: data.task_prompt,
    judgeRubric: data.judge_rubric,
    outputFormat: data.output_format ?? "free_text",
    judgeModel: data.judge_model ?? "claude-sonnet-4-20250514",
    difficultyTiers: data.difficulty_tiers ?? null,
    referenceContext: data.reference_context ?? null,
    referenceSources: data.reference_sources ?? null,
    requiredConcepts: data.required_concepts ?? null,
    calibrationExamples: data.calibration_examples ?? null,
    contextPreparation: data.context_preparation ?? null,
    requiredContextKeys: data.required_context_keys ?? null,
    maxRounds: data.max_rounds ?? 1,
    qualityThreshold: data.quality_threshold ?? 0.9,
    revisionPrompt: data.revision_prompt ?? null,
    sampleInput: data.sample_input ?? null,
  });
}
